test_utils.go

```diff
@@ -0,0 +1,151 @@
+package growthbook
+
+import (
+	"fmt"
+	"math"
+	"testing"
+)
+
+// Some test functions generate warnings in the log. We need to check
+// the expected ones, and not miss any unexpected ones.
+
+func handleExpectedWarnings(
+	t *testing.T, test []interface{}, expectedWarnings map[string]int) {
+	name, ok := test[0].(string)
+	if !ok {
+		t.Errorf("can't extract test name!")
+	}
+	warnings, ok := expectedWarnings[name]
+	if ok {
+		if len(testLog.errors) == 0 && len(testLog.warnings) == warnings {
+			testLog.reset()
+		} else {
+			t.Errorf("expected log warning")
+		}
+	}
+}
+
+// Helper to round variation ranges for comparison with fixed test
+// values.
+func roundRanges(ranges []Range) []Range {
+	result := make([]Range, len(ranges))
+	for i, r := range ranges {
+		rmin := math.Round(r.Min*1000000) / 1000000
+		rmax := math.Round(r.Max*1000000) / 1000000
+		result[i] = Range{rmin, rmax}
+	}
+	return result
+}
+
+// Helper to round floating point arrays for test comparison.
+func round(vals []float64) []float64 {
+	result := make([]float64, len(vals))
+	for i, v := range vals {
+		result[i] = math.Round(v*1000000) / 1000000
+	}
+	return result
+}
+
+// Logger to capture error and log messages.
+type testLogger struct {
+	errors   []string
+	warnings []string
+	info     []string
+}
+
+var testLog = testLogger{}
+
+func (log *testLogger) allErrors() string {
+	s := ""
+	for i, e := range log.errors {
+		if i != 0 {
+			s += ", "
+		}
+		s += e
+	}
+	return s
+}
+
+func (log *testLogger) allWarnings() string {
+	s := ""
+	for i, e := range log.warnings {
+		if i != 0 {
+			s += ", "
+		}
+		s += e
+	}
+	return s
+}
+
+func (log *testLogger) allInfo() string {
+	s := ""
+	for i, e := range log.info {
+		if i != 0 {
+			s += ", "
+		}
+		s += e
+	}
+	return s
+}
```

Comment:
Why manually implement joining with comma? The `strings.Join` method can support this with less code:

```go
func (log *testLogger) allErrors() string {
	return strings.Join(log.errors, ", ")
}
```